source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -49,25 +58,83 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
   lookup_ = cache_.makeLookupContext(std::move(lookup_request));
 
   ASSERT(lookup_);
-
   ENVOY_STREAM_LOG(debug, "CacheFilter::decodeHeaders starting lookup", *decoder_callbacks_);
-  lookup_->getHeaders([this](LookupResult&& result) { onHeaders(std::move(result)); });
-  if (state_ == GetHeadersState::GetHeadersResultUnusable) {
+
+  // Used exclusively to access the RequestHeaderMap in onHeaders
+  request_headers_ = &headers;
+  getHeaders();
+
+  if (get_headers_state_ == GetHeadersState::GetHeadersResultUnusable) {
     // onHeaders has already been called, and no usable cache entry was found--continue iteration.
     return Http::FilterHeadersStatus::Continue;
   }
   // onHeaders hasn't been called yet--stop iteration to wait for it, and tell it that we stopped
   // iteration.
-  state_ = GetHeadersState::FinishedGetHeadersCall;
+  get_headers_state_ = GetHeadersState::FinishedGetHeadersCall;
   return Http::FilterHeadersStatus::StopAllIterationAndWatermark;
 }
 
 Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
                                                      bool end_stream) {
-  // If lookup_ is null, the request wasn't cacheable, so the response isn't either.
-  if (lookup_ && request_allows_inserts_ && CacheabilityUtils::isCacheableResponse(headers)) {
-    // TODO(yosrym93): Add date internal header or metadata to cached responses and use it instead
-    // of the date header
+  if (skip_encoding_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  filter_state_ = FilterState::Encoding;
+
+  // If lookup_ is null, the request wasn't cacheable, so the response isn't either
+  if (!lookup_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (validating_cache_entry_ && isResponseNotModified(headers)) {
+    ASSERT(lookup_result_, "CacheFilter trying to validate a non-existent lookup result");
```

Comment:
The paths in the code ensure that `validating_cache_entry` is not set unless there is a valid `lookup_result_`.  My understanding is that these ASSERTs are used for development purposes (so that no one breaks any logic). This pattern has been already existing in the cache tree and I am just following it.
However, if the whole pattern is wrong maybe we should fix it? According to a previous discussion I had with @htuch in a previous PR this pattern of ASSERTs is okay.

Reply:
Specifically for the case of (non-RELEASE_) `ASSERT(!=nullptr)` right before a dereference, yes, I would say the whole pattern is wrong. You're going to crash regardless, so why not have a helpful fatal log message rather than a segfault? If you have a logical guarantee that it won't happen and want to cut down on `RELEASE_ASSERT`s, then the right choice would be to remove it entirely, rather than just dropping down to plain `ASSERT`. I can see using `ASSERT` as a sort of documentation of that assumption, but... again, it is impossible for `RELEASE_ASSERT` to cause trouble that wasn't already going to happen, so why not upgrade to it and get the nice error message?

Reply:
Oh wait, because ASSERT isn't compiled in, right? Ok, looking at it that way, ASSERT as documentation without going to RELEASE_ASSERT makes sense.

Reply:
That is my understanding ^